File: generator/generateImpulseMotion.m
function builtMotion = generateImpulseMotion(srcMotion)
    % GENERATEIMPULSEMOTION  Generate a built motion for an impulse primitive.
    %   builtMotion = generateImpulseMotion(srcMotion) takes a source motion struct
    %   and produces a built motion struct with a brief force spike at the start.
    %
    %   Input:
    %       srcMotion - Struct with id, name, motionRef, parameters (duration, direction).
    %   Output:
    %       builtMotion - Struct with id, name, compositionDegree, color, 
    %                     shortDisplayName, longDisplayName, magnitude, offset,
    %                     duration, flShape, frShape, rlShape, rrShape.

    % Extract parameters from source motion
    duration = srcMotion.parameters.duration;         % Seconds
    direction = srcMotion.parameters.direction;       % e.g., 'heave'
    magnitude = 300;                                  % Fixed magnitude of 300N

    % Constants
    Fs = 100;                                         % Sampling frequency (Hz)
    Ts = 1/Fs;                                        % Sampling period (0.01s)
    signalLen = round(duration * Fs);                 % Number of samples

    % Validate inputs
    if duration <= 0
        error('Duration must be positive');
    end

    % Generate impulse sequence: spike at t=0, then zero
    y = zeros(signalLen, 1);
    y(1) = 1;                                         % Single-sample impulse at start (normalized to 1)
    y = y * ones(1, 4);                              % Replicate for 4 actuators

    % Apply direction to adjust actuator sequences
    flShape = y(:, 1);
    frShape = y(:, 2);
    rlShape = y(:, 3);
    rrShape = y(:, 4);
    switch lower(direction)
        case 'heave'
            % All actuators same
        case 'pitch'
            rlShape = -rlShape;                       % Rear opposite
            rrShape = -rrShape;
        case 'roll'
            frShape = -frShape;                       % Right side opposite
            rrShape = -rrShape;
        case 'fl'
            frShape = zeros(signalLen, 1);            % Only FL active
            rlShape = zeros(signalLen, 1);
            rrShape = zeros(signalLen, 1);
        case 'fr'
            flShape = zeros(signalLen, 1);            % Only FR active
            rlShape = zeros(signalLen, 1);
            rrShape = zeros(signalLen, 1);
        case 'rl'
            flShape = zeros(signalLen, 1);            % Only RL active
            frShape = zeros(signalLen, 1);
            rrShape = zeros(signalLen, 1);
        case 'rr'
            flShape = zeros(signalLen, 1);            % Only RR active
            frShape = zeros(signalLen, 1);
            rlShape = zeros(signalLen, 1);
        otherwise
            error('Invalid direction: %s', direction);
    end

    % Construct built motion struct
    builtMotion = struct();
    builtMotion.id = srcMotion.id;
    builtMotion.name = srcMotion.name;
    builtMotion.compositionDegree = 0;                % Primitive motion
    builtMotion.color = helperGenerateRandomColor();  % Random hex color
    builtMotion.shortDisplayName = 'Impulse';
    builtMotion.longDisplayName = 'Impulse (300N)';
    builtMotion.magnitude = magnitude;                % Fixed at 300N
    builtMotion.offset = 0;                           % Default offset
    builtMotion.duration = duration;
    builtMotion.flShape = flShape;                    % [signalLen x 1], -1 to 1
    builtMotion.frShape = frShape;
    builtMotion.rlShape = rlShape;
    builtMotion.rrShape = rrShape;
end

File: generator/generate_impulse_motion.py
import numpy as np
from .helper_generate_random_color import helper_generate_random_color

def generate_impulse_motion(src_motion):
    """
    Generate a built motion with an impulse spike at the start.

    Parameters:
        src_motion (dict):
            - 'name'
            - 'parameters': {
                - 'duration' (float, seconds)
                - 'direction' (str, like 'heave')
                }

    Returns:
        dict: builtMotion with actuator waveforms and metadata.
    """
    # Extract parameters
    duration = src_motion['parameters']['duration']
    direction = src_motion['parameters']['direction'].lower()
    magnitude = 300  # Fixed value in Newtons

    # Constants
    Fs = 100
    Ts = 1 / Fs
    signal_len = round(duration * Fs)

    # Validation
    if duration <= 0:
        raise ValueError("Duration must be positive")

    # Generate impulse: 1 at t=0, then 0
    y = np.zeros((signal_len, 1))
    y[0, 0] = 1
    y = np.tile(y, (1, 4))  # replicate across 4 actuators

    # Apply direction shaping
    fl, fr, rl, rr = [y[:, i] for i in range(4)]
    if direction == 'heave':
        pass
    elif direction == 'pitch':
        rl *= -1
        rr *= -1
    elif direction == 'roll':
        fr *= -1
        rr *= -1
    elif direction == 'fl':
        fr[:] = rl[:] = rr[:] = 0
    elif direction == 'fr':
        fl[:] = rl[:] = rr[:] = 0
    elif direction == 'rl':
        fl[:] = fr[:] = rr[:] = 0
    elif direction == 'rr':
        fl[:] = fr[:] = rl[:] = 0
    elif direction == 'front':
        rl[:] = rr[:] = 0
    elif direction == 'rear':
        fl[:] = fr[:] = 0
    elif direction == 'left':
        fr[:] = rr[:] = 0
    elif direction == 'right':
        fl[:] = rl[:] = 0
    else:
        raise ValueError(f"Invalid direction: {direction}")

    # Build result
    built_motion = {
        'name': src_motion['name'],
        'compositionDegree': 0,
        'color': helper_generate_random_color(),
        'shortDisplayName': 'Impulse',
        'longDisplayName': 'Impulse (300N)',
        'magnitude': magnitude,
        'offset': 0,
        'duration': duration,
        'flShape': fl.tolist(),
        'frShape': fr.tolist(),
        'rlShape': rl.tolist(),
        'rrShape': rr.tolist(),
    }

    return built_motion
